transport/x/grpc/handler.go

```diff
@@ -0,0 +1,185 @@
+	return &handler{
+		procedureServiceName,
+		serviceName,
+		methodName,
+		router,
+		onewayErrorHandler,
+	}
+}
+
+func (h *handler) handle(
+	server interface{},
+	ctx context.Context,
+	decodeFunc func(interface{}) error,
+	interceptor grpc.UnaryServerInterceptor,
+) (interface{}, error) {
+	transportRequest, err := h.getTransportRequest(ctx, decodeFunc)
+	if err != nil {
+		return nil, err
+	}
+	if interceptor != nil {
+		return interceptor(
+			ctx,
+			transportRequest,
+			&grpc.UnaryServerInfo{
+				noopGrpcStruct{},
+				h.getFullMethod(),
+			},
+			func(ctx context.Context, request interface{}) (interface{}, error) {
+				transportRequest, ok := request.(*transport.Request)
+				if !ok {
+					return nil, fmt.Errorf("expected *transport.Request, got %T", request)
+				}
+				return h.call(ctx, transportRequest)
+			},
+		)
+	}
+	return h.call(ctx, transportRequest)
+}
+
+func (h *handler) getTransportRequest(ctx context.Context, decodeFunc func(interface{}) error) (*transport.Request, error) {
+	md, ok := metadata.FromContext(ctx)
+	if md == nil || !ok {
+		return nil, fmt.Errorf("cannot get metadata from ctx: %v", ctx)
+	}
+	caller, err := getCaller(md)
+	if err != nil {
+		return nil, err
+	}
+	if caller == "" {
+		caller = h.serviceName
+	}
+	encoding, err := getEncoding(md)
+	if err != nil {
+		return nil, err
+	}
+	if encoding == "" {
+		encoding = protobuf.Encoding
+	}
+	service, err := getService(md)
+	if err != nil {
+		return nil, err
+	}
+	if service == "" {
+		service = h.procedureServiceName
+	}
+	headers, err := getApplicationHeaders(md)
+	if err != nil {
+		return nil, err
+	}
+	var data []byte
+	if err := decodeFunc(&data); err != nil {
+		return nil, err
+	}
+	transportRequest := &transport.Request{
+		Caller:    caller,
+		Encoding:  encoding,
+		Service:   service,
+		Procedure: procedureToName(h.serviceName, h.methodName),
+		Headers:   headers,
+		Body:      bytes.NewBuffer(data),
+	}
+	if err := transport.ValidateRequest(transportRequest); err != nil {
+		return nil, err
+	}
+	return transportRequest, nil
+}
+
+func (h *handler) getFullMethod() string {
+	return fmt.Sprintf("/%s/%s", h.serviceName, h.methodName)
+}
+
+func (h *handler) call(ctx context.Context, transportRequest *transport.Request) (interface{}, error) {
+	handlerSpec, err := h.router.Choose(ctx, transportRequest)
+	if err != nil {
+		return nil, err
+	}
+	switch handlerSpec.Type() {
+	case transport.Unary:
+		return h.callUnary(ctx, transportRequest, handlerSpec.Unary())
+	case transport.Oneway:
+		return nil, h.callOneway(ctx, transportRequest, handlerSpec.Oneway())
+	default:
+		return nil, errors.UnsupportedTypeError{"grpc", handlerSpec.Type().String()}
+	}
+}
+
+func (h *handler) callUnary(ctx context.Context, transportRequest *transport.Request, unaryHandler transport.UnaryHandler) (interface{}, error) {
+	if err := request.ValidateUnaryContext(ctx); err != nil {
+		return nil, err
+	}
+	protobuf.SetRawResponse(transportRequest.Headers)
```

Comment:
I've added a comment about this. Thanks for the catch on setting the value, this is per the documentation of `transport.Headers#With` that says you cannot rely on the input object being the same as the return object, I've updated the documenation on `protobuf.SetRawResponse` as well.